src/gbinder_eventloop.c

```diff
@@ -0,0 +1,322 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "gbinder_eventloop_p.h"
+
+#include <gutil_macros.h>
+
+#define GBINDER_DEFAULT_EVENTLOOP (&gbinder_eventloop_glib)
+
+static const GBinderEventLoopIntegration gbinder_eventloop_glib;
+
+/*==========================================================================*
+ * GLib event loop integration
+ *==========================================================================*/
+
+typedef struct gbinder_eventloop_glib_timeout {
+    GBinderEventLoopTimeout timeout;
+    guint id;
+    GSourceFunc func;
+    gpointer data;
+} GBinderEventLoopTimeoutGLib;
+
+typedef struct gbinder_eventloop_glib_callback {
+    GSource source;
+    GBinderEventLoopCallback callback;
+} GBinderEventLoopCallbackGLib;
+
+static
+inline
+GBinderEventLoopTimeoutGLib*
+gbinder_eventloop_glib_timeout_cast(
+    GBinderEventLoopTimeout* timeout)
+{
+    return G_CAST(timeout,GBinderEventLoopTimeoutGLib,timeout);
+}
+
+static
+inline
+GSource*
+gbinder_eventloop_glib_callback_source(
+    GBinderEventLoopCallback* callback)
+{
+    return &(G_CAST(callback,GBinderEventLoopCallbackGLib,callback)->source);
+}
+
+static
+gboolean
+gbinder_eventloop_glib_timeout_callback(
+    gpointer data)
+{
+    GBinderEventLoopTimeoutGLib* timeout = data;
+
+    return timeout->func(timeout->data);
+}
+
+static
+void
+gbinder_eventloop_glib_timeout_finalize(
+    gpointer data)
+{
+    g_slice_free1(sizeof(GBinderEventLoopTimeoutGLib), data);
+}
+
+static
+GBinderEventLoopTimeout*
+gbinder_eventloop_glib_timeout_add(
+    guint interval,
+    GSourceFunc func,
+    gpointer data)
+{
+    GBinderEventLoopTimeoutGLib* impl =
+        g_slice_new(GBinderEventLoopTimeoutGLib);
+
+    impl->timeout.eventloop = &gbinder_eventloop_glib;
+    impl->func = func;
+    impl->data = data;
+    impl->id = g_timeout_add_full(G_PRIORITY_DEFAULT, interval,
+        gbinder_eventloop_glib_timeout_callback, impl,
+        gbinder_eventloop_glib_timeout_finalize);
+    return &impl->timeout;
+}
+
+static
+void
+gbinder_eventloop_glib_timeout_remove(
+    GBinderEventLoopTimeout* timeout)
+{
+    g_source_remove(gbinder_eventloop_glib_timeout_cast(timeout)->id);
+}
+
+static
+gboolean
+gbinder_eventloop_glib_callback_prepare(
+    GSource* source,
+    gint* timeout)
+{
+    *timeout = 0;
+    return TRUE;
+}
+
+static
+gboolean
+gbinder_eventloop_glib_callback_check(
+    GSource* source)
+{
+    return TRUE;
+}
+
+static
+gboolean
+gbinder_eventloop_glib_callback_dispatch(
+    GSource* source,
+    GSourceFunc callback,
+    gpointer user_data)
+{
+    ((GBinderEventLoopCallbackFunc)callback)(user_data);
+    return G_SOURCE_REMOVE;
```

Comment:
This function casting looks ugly. Claiming glib convention is followed (cb return boolean keep/remove info) then ignoring it? Why / what is going on?

Reply:
GLib does this all the time :) That simplifies the implementation (GSource already provides storage for the callback pointer, why not to use it), interface is clean and type-safe, implementation can be re-written if needed. Here I essentially traded pointer-sized amount of memory for a cast.

Reply:
glib code does make all kinds of crazy stuff all the time - like casting function pointers which invites undefined behavior even if it happens to workfor<xxx>...

From c99 draft spec that I happen to have available:

6.3.2.3 Pointers

"A pointer to a function of one type may be converted to a pointer to a function of another
type and back again; the result shall compare equal to the original pointer. If a converted
pointer is used to call a function whose type is not compatible with the pointed-to type,
the behavior is undefined."

Reply:
Then again, in the absence of things like mixed calling conventions and whatnot it is probably going to workforus at least as long as glib does.

Reply:
Except when - say - function that uses pascal calling convention gets treated as one that uses c calling convention i.e. crash boom due to incorrect assumption about whether callee or caller should clean up the stack frame...

Reply:
... also in such cases args are pushed to stack in different order.